File: Java/Striver/sorting_I.java
public class sorting_I {

    public static void merge(int nums[]) {
        if (nums.length < 2)
            return;
        int left = nums.length / 2;
        int right = nums.length - nums.length / 2;

        int L[] = new int[left];
        int R[] = new int[right];

        for (int i = 0; i < left; i++) {
            L[i] = nums[i];
        }
        for (var j = 0; j < right; j++) {
            R[j] = nums[left + j];
        }

        merge(L);
        merge(R);
        sort(nums, L, R);

    }

    public static void sort(int nums[], int left[], int right[]) {
        int i = 0, j = 0, k = 0;
        while (i < left.length && j < right.length) {
            if (left[i] < right[j]) {
                nums[k] = left[i];
                i++;
            } else {
                nums[k] = right[j];
                j++;
            }
            k++;
        }

        while (i < left.length) {
            nums[k] = left[i];
            i++;
            k++;
        }

        while (j < right.length) {
            nums[k] = right[j];
            j++;
            k++;
        }
    }

    public static void quickSort(int[] nums) {

    }

    public static void recurBubbleSort(int[] nums, int n) {
        // Base case: range == 1
        if (n == 1)
            return;

        int didSwap = 0;
        for (int i = 0; i <= n - 2; i++) {
            if (nums[i] > nums[i + 1]) {
                int temp = nums[i];
                nums[i] = nums[i + 1];
                nums[i + 1] = temp;
                didSwap = 1;
            }

        }
        if (didSwap == 0) {
            return;
        }

        recurBubbleSort(nums, n - 1);

    }

    public static void InsertionSort(int[] nums) {
        for (int i = 0; i < nums.length; i++) {
            int j = i;
            while (j > 0 && nums[j - 1] > nums[j]) {
                int temp = nums[j - 1];
                nums[j - 1] = nums[j];
                nums[j] = temp;
                j--;
            }
        }
        System.out.println("After using insertion sort: ");
        for (var i = 0; i < nums.length; i++) {
            System.out.print(nums[i] + " ");
        }
        System.out.println();
    }

    public static void recurInsertionSort(int nums[], int i, int n) {
        if (i == n) {
            return;
        }
        int j = i;
        while (j > 0 && nums[j - 1] > nums[j]) {
            int temp = nums[j];
            nums[j] = nums[j - 1];
            nums[j - 1] = temp;
            j--;
        }
        recurInsertionSort(nums, i + 1, n);
    }

    public static void findLargest(int nums[]) {
        int max = nums[0];
        for (int i = 0; i < nums.length; i++) {
            if (nums[i] > max) {
                max = nums[i];
            }
        }

        System.out.println(max);
    }

    public static boolean CheckSorted(int nums[]) {
        for (int i = 0; i < nums.length - 1; i++) {
            if (nums[i] > nums[i + 1]) {
                return false;
            }
        }

        return true;
    }

    public static void main(String args[]) {
        int[] nums = { 1, 23, 4, 5, 4 };
      

  
        System.out.println();
    }

}

File: cpp/Striver/Sorting_I/Sorting_I.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution
{
public:
    void selectionSort(int nums[], int n)
    {
        // selection sort
        for (int i = 0; i < n - 1; i++)
        {
            int min = i;
            for (int j = i + 1; j < n; j++)
            {
                if (nums[j] < nums[min])
                {
                    min = j;
                }
            }
            int temp = nums[min];
            nums[min] = nums[i];
            nums[i] = temp;
        }
        cout << "After the selection sort: "
             << "\n";
        for (int i = 0; i < n; i++)
        {
            cout << nums[i] << " ";
        }
        cout << endl;
    }

    void BubbleSort1(int nums[], int n)
    {
        // Bubble sort
        for (int i = n - 1; i >= 0; i--)
        {
            int didSwap = 0;
            for (int j = 0; j <= i - 1; j++)
            {
                if (nums[j] > nums[j + 1])
                {
                    int temp = nums[j + 1];
                    nums[j + 1] = nums[j];
                    nums[j] = temp;
                    didSwap = 1;
                }
            }

            if (didSwap == 0)
            {
                break;
            }
        }
        cout << "After Using bubble sort : "
             << "\n";

        for (int i = 0; i < n; i++)
        {
            cout << nums[i] << endl;
        }
        cout << "\n";
    }
    void BubbleSort2(int nums[], int n)
    {
        // bubble sort
        for (int i = n - 1; i >= 0; i--)
        {
            int didSwap = 0;
            for (int j = 0; j <= i - 1; j++)
            {
                if (nums[j] > nums[j + 1])
                {
                    int temp = nums[j + 1];
                    nums[j + 1] = nums[j];
                    nums[j] = temp;
                    didSwap = 1;
                }
            }
            if (didSwap == 0)
            {
                break;
            }
        }

        cout << "After using bubble sort: " << endl;
        for (int i = 0; i < n; i++)
        {
            cout << nums[i] << " ";
        }
        cout << endl;
    }

    void insertion_sort(int nums[], int n)
    {
        for (int i = 0; i < n; i++)
        {
            int j = i;
            while (j > 0 && nums[j - 1] > nums[j])
            {
                int temp = nums[j - 1];
                nums[j - 1] = nums[j];
                nums[j] = temp;
                j--;
            }
        }
        cout << "After using insertion sort : " << endl;
        for (int i = 0; i < n; i++)
        {
            cout << nums[i] << " ";
        }
        cout << endl;
    }
    int sortArr(vector<int> &arr)
    {
        sort(arr.begin(), arr.end());
        return arr[arr.size() - 1];
    }

    // int findLargestElement(int arr[])
    //{
    //     int max = arr[0];
    //     for (int i = 0; i < sizeof(arr) / sizeof(int); i++)
    //     {
    //         if (arr[i] > max)
    //         {
    //             max = arr[i];
    //         }
    //     }
    //     cout << max;
    //     return max;
    // }

    void insertionSort(vector<int> &nums)
    {
        for (int i = 0; i < nums.size(); i++)
        {
            int j = i;
            while (j > 0 && nums[j - 1] > nums[j])
            {
                int temp = nums[j - 1];
                nums[j - 1] = nums[j];
                nums[j] = temp;
                j--;
            };
        }
    }

    void BubbleSort3(vector<int> &nums)
    {
        bool swapped;
        for (int i = nums.size() - 1; i >= 0; i--)
        {
            swapped = 0;
            for (int j = 0; j < i; j++)
            {
                if (nums[j] > nums[j + 1])
                {
                    int temp = nums[j];
                    nums[j] = nums[j + 1];
                    nums[j + 1] = temp;
                    swapped = 1;
                }
                if (swapped == 0)
                {
                    break;
                }
            }
        }
    }

    void insrtnSort(vector<int> &nums, int i, int n)
    {
        if (i == n)
            return;

        int j = i;
        while (j > 0 && nums[j - 1] > nums[j])
        {
            int temp = nums[j];
            nums[j] = nums[j - 1];
            nums[j - 1] = temp;
            j--;
        }
        insrtnSort(nums, i + 1, n);
    }
};

int main()
{
    vector<int> nums = {5, 4, 3, 2, 1};
    Solution sol;
    sol.insrtnSort(nums, 0, nums.size());
    for (auto &a : nums)
    {
        cout << a << " ";
    }
    cout << endl;
    return 1;
}
